src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -2006,3 +2055,458 @@ ShouldInitiateMetadataSync(bool *lockFailure)
+	}
+}
+
+
+/*
+ * EnsureShardMetadataIsSane ensures that the input values are safe
+ * for inserting into pg_dist_shard metadata.
+ */
+static void
+EnsureShardMetadataIsSane(Oid relationId, int64 shardId, char storageType,
+						  text *shardMinValue, text *shardMaxValue)
+{
+	if (shardId <= INVALID_SHARD_ID)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Invalid shard id: %ld", shardId)));
+	}
+
+	if (!(storageType == SHARD_STORAGE_TABLE ||
+		  storageType == SHARD_STORAGE_FOREIGN ||
+		  storageType == SHARD_STORAGE_COLUMNAR))
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Invalid shard storage type: %c", storageType)));
+	}
+
+	char partitionMethod = PartitionMethodViaCatalog(relationId);
+	if (partitionMethod == DISTRIBUTE_BY_INVALID)
+	{
+		/* connection from the coordinator operating on a shard */
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("The relation \"%s\" does not have a valid "
+							   "entry in pg_dist_partition.",
+							   get_rel_name(relationId))));
+	}
+	else if (!(partitionMethod == DISTRIBUTE_BY_HASH ||
+			   partitionMethod == DISTRIBUTE_BY_NONE))
+	{
+		/* connection from the coordinator operating on a shard */
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Metadata syncing is only allowed for hash, "
+							   "reference and local tables: %c", partitionMethod)));
+	}
+
+	List *distShardTupleList = LookupDistShardTuples(relationId);
+	if (partitionMethod == DISTRIBUTE_BY_NONE)
+	{
+		if (!(shardMinValue == NULL && shardMaxValue == NULL))
+		{
+			char *relationName = get_rel_name(relationId);
+			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+							errmsg("Shards of reference or local table \"%s\" should "
+								   "have NULL shard ranges", relationName)));
+		}
+		else if (list_length(distShardTupleList) != 0)
+		{
+			char *relationName = get_rel_name(relationId);
+			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+							errmsg("relation \"%s\" has already at least one shard, "
+								   "adding more is not allowed", relationName)));
+		}
+	}
+	else if (partitionMethod == DISTRIBUTE_BY_HASH)
+	{
+		if (shardMinValue == NULL || shardMaxValue == NULL)
+		{
+			char *relationName = get_rel_name(relationId);
+			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+							errmsg("Shards of has distributed table  \"%s\" "
+								   "cannot have NULL shard ranges", relationName)));
+		}
+
+		char *shardMinValueString = text_to_cstring(shardMinValue);
+		char *shardMaxValueString = text_to_cstring(shardMaxValue);
+
+		/* pg_strtoint32 does the syntax and out of bound checks for us */
+		int32 shardMinValueInt = pg_strtoint32(shardMinValueString);
+		int32 shardMaxValueInt = pg_strtoint32(shardMaxValueString);
+
+		if (shardMinValueInt >= shardMaxValueInt)
```

Comment:
I think them being equal should be allowed right? It would only store one hash value in that case. Or is shardMaxValueInt an exclusive bound?

```suggestion
		if (shardMinValueInt > shardMaxValueInt)
```

Reply:
`shardMaxValueInt ` is inclusive. But I'm actually not sure if we can ever have shard min == shard max, and we need 4b shards for that. To prevent malicious users doing weird things, I suggest we keep as is for now

Reply:
doesn't tenant isolation do exactly this?

Reply:
ops, I think yes. Let me add tenant isolation test as well, thanks for reminding